packages/tracing/src/browser/request.ts

```diff
@@ -0,0 +1,234 @@
+import { addInstrumentationHandler, isInstanceOf, isMatchingPattern } from '@sentry/utils';
+
+import { Span } from '../span';
+
+import { getActiveTransaction } from './utils';
+
+export const DEFAULT_TRACING_ORIGINS = ['localhost', /^\//];
+
+/** Options for Request Instrumentation */
+export interface RequestInstrumentationOptions {
+  /**
+   * List of strings / regex where the integration should create Spans out of. Additionally this will be used
+   * to define which outgoing requests the `sentry-trace` header will be attached to.
+   *
+   * Default: ['localhost', /^\//] {@see DEFAULT_TRACING_ORIGINS}
+   */
+  tracingOrigins: Array<string | RegExp>;
+
+  /**
+   * Flag to disable patching all together for fetch requests.
+   *
+   * Default: true
+   */
+  traceFetch: boolean;
+
+  /**
+   * Flag to disable patching all together for xhr requests.
+   *
+   * Default: true
+   */
+  traceXHR: boolean;
+
+  /**
+   * This function will be called before creating a span for a request with the given url.
+   * Return false if you don't want a span for the given url.
+   *
+   * By default it uses the `tracingOrigins` options as a url match.
+   */
+  shouldCreateSpanForRequest?(url: string): boolean;
+}
+
+/** Data returned from fetch callback */
+interface FetchData {
+  args: any[];
+  fetchData: {
+    method: string;
+    url: string;
+    // span_id
+    __span?: string;
+  };
+  startTimestamp: number;
+  endTimestamp?: number;
+}
+
+/** Data returned from XHR request */
+interface XHRData {
+  xhr?: {
+    __sentry_xhr__?: {
+      method: string;
+      url: string;
+      status_code: number;
+      data: Record<string, any>;
+    };
+    __sentry_xhr_span_id__?: string;
+    __sentry_own_request__: boolean;
+    setRequestHeader?: Function;
+  };
+  startTimestamp: number;
+  endTimestamp?: number;
+}
+
+export const defaultRequestInstrumentionOptions: RequestInstrumentationOptions = {
+  traceFetch: true,
+  traceXHR: true,
+  tracingOrigins: DEFAULT_TRACING_ORIGINS,
+};
+
+/** Registers span creators for xhr and fetch requests  */
+export function registerRequestInstrumentation(_options?: Partial<RequestInstrumentationOptions>): void {
+  const { traceFetch, traceXHR, tracingOrigins, shouldCreateSpanForRequest } = {
+    ...defaultRequestInstrumentionOptions,
+    ..._options,
+  };
+
+  // We should cache url -> decision so that we don't have to compute
+  // regexp everytime we create a request.
+  const urlMap: Record<string, boolean> = {};
+
+  const defaultShouldCreateSpan = (url: string): boolean => {
+    if (urlMap[url]) {
+      return urlMap[url];
+    }
+    const origins = tracingOrigins;
+    urlMap[url] =
+      origins.some((origin: string | RegExp) => isMatchingPattern(url, origin)) &&
+      !isMatchingPattern(url, 'sentry_key');
+    return urlMap[url];
+  };
+
+  const shouldCreateSpan = shouldCreateSpanForRequest || defaultShouldCreateSpan;
+
+  const spans: Record<string, Span | undefined> = {};
```

Comment:
I am not sure if that is the best approach here, we never ever remove a span from here, right?
So at one point, the user will run out of memory because we don't empty this after sending it.

This is probably tricky to solve since we can't just delete spans that are finished, we need to wait until the Transaction containing the span finished.

Reply:
We can just delete the span from the map on finish. Good point.